pkg/controllers/state/suite_test.go

```diff
@@ -779,6 +801,45 @@ var _ = Describe("Inflight Nodes", func() {
 			},
 		}))
 	})
+	It("should only set startup taints once", func() {
+		taints := []v1.Taint{
+			{
+				Key:    "custom-taint1",
+				Value:  "custom-value1",
+				Effect: v1.TaintEffectNoSchedule,
+			},
+		}
+
+		customProvisioner := test.Provisioner(test.ProvisionerOptions{
+			ObjectMeta:    metav1.ObjectMeta{Name: "custom-provisioner"},
+			StartupTaints: taints,
+		})
+		ExpectApplied(ctx, env.Client, customProvisioner)
+
+		node := test.Node(test.NodeOptions{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "abcd",
+				Labels: map[string]string{
+					v1alpha5.ProvisionerNameLabelKey: customProvisioner.Name,
+					v1.LabelInstanceTypeStable:       cloudProvider.InstanceTypes[0].Name,
+				},
+			},
+			Taints:     taints,
+			ProviderID: test.RandomProviderID(),
+		})
+		ExpectApplied(ctx, env.Client, node)
+		ExpectReconcileSucceeded(ctx, nodeController, client.ObjectKeyFromObject(node))
+
+		// Update the provisioner to the default suite provisioner w/ no StartupTaints. StartupTaints should still be
+		// populated by custom provisioner.
+		node.Labels[v1alpha5.ProvisionerNameLabelKey] = provisioner.Name
+		ExpectApplied(ctx, env.Client, node)
+		ExpectReconcileSucceeded(ctx, nodeController, client.ObjectKeyFromObject(node))
+
+		ExpectStateNodeCount("==", 1)
+		stateNode := ExpectStateNodeExists(node)
+		Expect(stateNode.Taints()).To(HaveLen(0))
```

Comment:
Shouldn't this be len(1)? The taints are added, and on the second pass, it should just propagate what the old state node got.

Reply:
`Taints()` returns an empty slice because it's using the list of startup taints from the `customProvisioner`. Since the node is not yet initialized it filters the startup taints out of the taint list. If the node were initialized we would see the taints.